assertsprocessor/sampler.go

```diff
@@ -51,68 +52,91 @@ func (s *sampler) stopProcessing() {
 	s.stopTraceFlusher()
 }
 
-func (s *sampler) sampleTraces(ctx context.Context, traces *resourceTraces) {
-	for _, traceStruct := range *traces.traceById {
-		if traceStruct.getMainSpan() == nil {
-			continue
+func (s *sampler) sampleTraces(ctx context.Context, traces []*trace) {
+	for _, tr := range traces {
+		sampled := false
+		for _, ts := range tr.segments {
+			if ts.getMainSpan() == nil {
+				continue
+			}
+			s.updateTrace(ts.namespace, ts.service, ts)
+			entityKeyString := ts.requestKey.entityKey.AsString()
+
+			// Get the trace queue for the entity and request
+			perService, _ := s.topTracesByService.LoadOrStore(entityKeyString, newServiceQueues(s.config))
+			request := ts.requestKey.request
+			requestState := perService.(*serviceQueues).getRequestState(request)
+			if requestState == nil {
+				s.logger.Warn("Too many requests in Entity. Dropping",
+					zap.String("Entity", entityKeyString),
+					zap.String("Request", request))
+				return
+			}
+
+			item := Item{
+				trace:   tr,
+				ctx:     &ctx,
+				latency: ts.latency,
+			}
+			s.incrTotalTraceCount(ts)
+			for _, span := range ts.getSpans() {
+				if spanHasError(span) {
+					s.logger.Debug("Capturing error trace",
+						zap.String("traceId", span.TraceID().String()),
+						zap.String("service", entityKeyString),
+						zap.String("request", request),
+						zap.Float64("latency", ts.latency))
+					span.Attributes().PutStr(AssertsTraceSampleTypeAttribute, AssertsTraceSampleTypeError)
+
+					if !sampled {
+						s.incrSampledTraceCount(AssertsTraceSampleTypeError, ts)
+						requestState.errorQueue.push(&item)
+						sampled = true
+					}
+				} else if s.spanIsSlow(span, ts) {
+					s.logger.Debug("Capturing slow trace",
+						zap.String("traceId", span.TraceID().String()),
+						zap.String("service", entityKeyString),
+						zap.String("request", request),
+						zap.Float64("latency", ts.latency))
+					span.Attributes().PutStr(AssertsTraceSampleTypeAttribute, AssertsTraceSampleTypeSlow)
+
+					if !sampled {
+						s.incrSampledTraceCount(AssertsTraceSampleTypeSlow, ts)
+						requestState.slowQueue.push(&item)
+						sampled = true
+					}
+				}
+			}
 		}
-		s.updateTrace(traces.namespace, traces.service, traceStruct)
-		entityKeyString := traceStruct.requestKey.entityKey.AsString()
-
-		// Get the traceStruct queue for the entity and request
-		perService, _ := s.topTracesByService.LoadOrStore(entityKeyString, newServiceQueues(s.config))
-		request := traceStruct.requestKey.request
-		requestState := perService.(*serviceQueues).getRequestState(request)
-		if requestState == nil {
-			s.logger.Warn("Too many requests in Entity. Dropping",
-				zap.String("Entity", entityKeyString),
-				zap.String("Request", request))
-			return
+		if !sampled {
+			s.captureNormalTraceSample(ctx, tr)
```

Comment:
Shouldn't `sampled = true` be set here?

Reply:
This is outside the traceSegment for loop. We need it inside the traceSegment for loop to ensure we don't end up adding the same trace twice to avoid the duplicate segments in the trace backend.